### PR TITLE
DataGrid.Pdf - Indents for grouped rows

### PR DESCRIPTION
**Pull request by KuznetsovVN**

### CODE REVIEW

File: js/exporter/jspdf/draw_pdf_table.js [hunk 1]
```diff
@@ -49,6 +49,10 @@ export function drawPdfTable(doc, table) {
             if(!isDefined(cell._rect)) {
                 throw 'cell._rect is required';
             }
+            if(isDefined(cell.backgroundColor)) {
+                doc.setFillColor(cell.backgroundColor);
```
Comment (IgnatovDan) [inline js/exporter/jspdf/draw_pdf_table.js]: Let's move 'cell background' to a separate PR

File: js/exporter/jspdf/export_data_grid_2.js [hunk 2]
```diff
@@ -11,27 +11,52 @@ function exportDataGrid(doc, dataGrid, options) {
...
+            const rowsIndents = [];
 
             pdfGrid.startNewTable(options.drawTableBorder, options.topLeft);
 
             const dataRowsCount = dataProvider.getRowsCount();
 
             for(let rowIndex = 0; rowIndex < dataRowsCount; rowIndex++) {
                 const currentRow = [];
+                let groupLevel = 0;
+                let rowType;
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_data_grid_2.js]: let -> const?

File: js/exporter/jspdf/export_data_grid_2.js [hunk 3]
```diff
@@ -11,27 +11,52 @@ function exportDataGrid(doc, dataGrid, options) {
...
+                    if(rowType === 'header') {
                         const cellMerging = dataProvider.getCellMerging(rowIndex, cellIndex);
                         if(cellMerging && cellMerging.rowspan > 0) {
                             pdfCell.rowSpan = cellMerging.rowspan;
                         }
                         if(cellMerging && cellMerging.colspan > 0) {
                             pdfCell.colSpan = cellMerging.colspan;
                         }
+                    } else if(rowType === 'group') {
+                        const prevCellsHaveSummaryItems = currentRow.slice(1).filter(pdfCell => isDefined(pdfCell.text)).length > 0;
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_data_grid_2.js]: Let's move 'group totals' to a separate PR

File: js/exporter/jspdf/export_data_grid_2.js [hunk 4]
```diff
@@ -40,20 +65,35 @@ function exportDataGrid(doc, dataGrid, options) {
                     currentRow.push(pdfCell);
                 }
 
+                rowsIndents.push(groupLevel * 10); // TODO: Default value for horizontalIndent : 10
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_data_grid_2.js]: Let's add 'indent' into 'options' with 10 as the default value

File: js/exporter/jspdf/export_data_grid_2.js [hunk 5]
```diff
@@ -40,20 +65,35 @@ function exportDataGrid(doc, dataGrid, options) {
...
+                let startNewTableWithIndent = rowsIndents.length >= 2 && rowsIndents[rowsIndents.length - 1] !== rowsIndents[rowsIndents.length - 2];
+
                 let rowHeight = null; // TODO: Default Value
                 if(options.onRowExporting) {
                     const args = { drawNewTableFromThisRow: {}, rowCells: currentRow };
                     options.onRowExporting(args);
                     const { startNewTable, addPage, tableTopLeft, splitToTablesByColumns } = args.drawNewTableFromThisRow;
                     if(startNewTable === true) {
                         pdfGrid.startNewTable(options.drawTableBorder, tableTopLeft, addPage === true, splitToTablesByColumns);
+                        startNewTableWithIndent = false;
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_data_grid_2.js]: Let's move 'support group indents between pages' to a separate PR

File: js/exporter/jspdf/pdf_grid.js [hunk 6]
```diff
@@ -14,10 +14,13 @@ export class PdfGrid {
         this._newPageTables.push(this._currentHorizontalTables[this._currentHorizontalTables.length - 1]);
     }
 
-    startNewTable(drawTableBorder, firstTableTopLeft, firstTableOnNewPage, splitByColumns) {
+    startNewTable(drawTableBorder, firstTableTopLeft, firstTableOnNewPage, splitByColumns, firstColumnWidth) {
         if(isDefined(splitByColumns)) {
             this._splitByColumns = splitByColumns;
         }
+        if(isDefined(firstColumnWidth)) {
+            this._columnWidths[0] = firstColumnWidth;
```
Comment (IgnatovDan) [inline js/exporter/jspdf/pdf_grid.js]: Is it possible to avoid changes in _columnWidths?

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: js/exporter/jspdf/pdf_grid.js [hunk 7]
```diff
@@ -14,14 +14,14 @@ export class PdfGrid {
...
-    startNewTable(drawTableBorder, firstTableTopLeft, firstTableOnNewPage, splitByColumns) {
+    startNewTable(drawTableBorder, firstTableTopLeft, firstTableOnNewPage, splitByColumns, firstColumnWidth) {
         if(isDefined(splitByColumns)) {
             this._splitByColumns = splitByColumns;
         }
 
         const firstTableEndColumnIndex = this._splitByColumns[0]?.columnIndex ?? this._columnWidths.length;
         this._currentHorizontalTables = [
-            new PdfTable(drawTableBorder, firstTableTopLeft, this._columnWidths.slice(0, firstTableEndColumnIndex))
+            new PdfTable(drawTableBorder, firstTableTopLeft, this._columnWidths.slice(0, firstTableEndColumnIndex), firstColumnWidth)
```
Comment (IgnatovDan) [inline js/exporter/jspdf/pdf_grid.js]: const tableColumnWidths = this._columnWidths.slice(0, firstTableEndColumnIndex);
tableColumnWidths[0] = firstColumnWidth;
this._currentHorizontalTables = [new PdfTable(drawTableBorder, firstTableTopLeft, tableColumnWidths)];

?

File: js/exporter/jspdf/pdf_table.js [hunk 8]
```diff
@@ -11,13 +11,16 @@ export class PdfTable {
         }
 
         this.drawTableBorder = drawTableBorder;
+        this.columnWidths = columnWidths.slice(0); // TODO
```
Comment (IgnatovDan) [inline js/exporter/jspdf/pdf_table.js]: Implement before "new PdfTable" ?

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

**Review by novsstation — DISMISSED** [review]

File: js/exporter/jspdf/export_data_grid_2.js [hunk 9]
```diff
@@ -6,32 +6,40 @@ function exportDataGrid(doc, dataGrid, options) {
     if(!isDefined(options.topLeft)) {
         throw 'options.topLeft is required';
     }
+    options.indent = options.indent || 10;
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_data_grid_2.js]: Imho Number(0) should be available:
`isDefined(options.indent) ? options.indent : 10; `
Initialize it in _getFullOptions?

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

**Review by IgnatovDan — APPROVED** [review]

**Review by novsstation — APPROVED** [review]

**Review by EugeniyKiyashko — APPROVED** [review]

**Pull request merged**